bindings/rust/wasmedge-sys/src/vm.rs

```diff
@@ -0,0 +1,174 @@
+use super::wasmedge;
+use crate::{
+    raw_result::{decode_result, ErrReport},
+    string::StringRef,
+    value::Value,
+    wasi,
+};
+use std::os::raw::c_char;
+
+// If there is a third-party sdk based on wasmedge-sys
+// the private encapsulation  here can force the third-party sdk to use the api
+#[derive(Debug)]
+pub struct Vm {
+    pub(crate) ctx: *mut wasmedge::WasmEdge_VMContext,
+    _private: (),
+}
+
+impl Vm {
+    pub fn create(config: &crate::config::Config) -> Result<Self, ErrReport> {
+        let ctx = unsafe {
+            wasmedge::WasmEdge_VMCreate(config.ctx, std::ptr::null_mut() /* store */)
+        };
+        if ctx.is_null() {
+            panic!("WasmEdge VM create failed");
+        }
+        Ok(Self { ctx, _private: () })
+    }
+
+    pub fn load_wasm_from_ast_module(
+        self,
+        module: &crate::module::Module,
+    ) -> Result<Self, ErrReport> {
+        unsafe {
+            decode_result(wasmedge::WasmEdge_VMLoadWasmFromASTModule(
+                self.ctx, module.ctx,
+            ))?;
+        }
+        Ok(self)
+    }
+
+    pub fn validate(self) -> Result<Self, ErrReport> {
+        unsafe {
+            decode_result(wasmedge::WasmEdge_VMValidate(self.ctx))?;
+        }
+        Ok(self)
+    }
+
+    pub fn instantiate(self) -> Result<Self, ErrReport> {
+        unsafe {
+            decode_result(wasmedge::WasmEdge_VMInstantiate(self.ctx))?;
+        }
+        Ok(self)
+    }
+
+    pub fn init_wasi_obj(
+        &self,
+        args: Option<Vec<&str>>,
+        envs: Option<Vec<&str>>,
+        dirs: Option<Vec<&str>>,
+        preopens: Option<Vec<&str>>,
+    ) {
+        let import_mod_ctx = unsafe {
+            wasmedge::WasmEdge_VMGetImportModuleContext(
+                self.ctx,
+                wasmedge::WasmEdge_HostRegistration_Wasi,
+            )
+        };
+
+        let mut import_obj = wasi::ImportObjInitParams::new();
+
+        if let Some(mut args) = args {
+            import_obj.args = args
+                .iter_mut()
+                .map(|arg| {
+                    let raw_arg: wasmedge::WasmEdge_String = StringRef::from(*arg).into();
+                    raw_arg.Buf
+                })
+                .collect::<Vec<*const c_char>>();
+            import_obj.args_len = args.len() as u32;
+        }
+
+        if let Some(mut envs) = envs {
+            import_obj.envs = envs
+                .iter_mut()
+                .map(|env| {
+                    let raw_env: wasmedge::WasmEdge_String = StringRef::from(*env).into();
```

Comment:
can you run clippy on this? I don't think you need `mut`

Reply:
clippy is not foolproof, I execute it a many times and it won't prompt this.